### PR TITLE
Consolidate content versioning

### PR DESCRIPTION
**Pull request by licitdev**
## Scope

What's changed:

- Added a new `delta` field to the `directus_versions` table
  - Contains the final delta which is the combination of all revisions

## Potential Risks / Drawbacks

- Updates to existing revisions during the migration may result in an incorrect final delta stored
  - This happens when an existing instance of Directus is performing the patch
  - The final delta could be fully repopulated in a subsequent release

## Review Notes / Questions

- The existing `versions` field in the `directus_revisions` table will be removed in a subsequent release
- If the `delta` field in `directus_versions` is empty, the existing behaviour of fetching from `directus_revisions` remains
- Needs further testing to ensure content versioning remains fully operational

---

Related to #17894

### CODE REVIEW

File: api/src/database/migrations/20240415A-consolidate-content-versioning.ts
Comment (paescuj) [inline api/src/database/migrations/20240415A-consolidate-content-versioning.ts:1]: For instances with large number of Content Versions, it might make sense to process data in batches, similar to https://github.com/directus/directus/blob/4b5a03caf4fa4de579196bfe01bdf944defe8408/api/src/database/migrations/20240328A-permissions-policies.ts.

Also for such instances, do we have to worry about storage when duplicating the Content Versions or do you think this is negligible?

Comment (licitdev) [reply api/src/database/migrations/20240415A-consolidate-content-versioning.ts:1]: The database updates are done individually per version, so I think the `processChunk()` is not applicable as it's not a bulk operation that benefits from batches. 🤔 Do you envision this migration to be done in a different approach?

The impact should be minimal as only the final `delta` is added. This is added as part of the bigger improvement to clean up the activity & revisions table, by removing the dependency on those tables.

Comment (paescuj) [reply api/src/database/migrations/20240415A-consolidate-content-versioning.ts:1]: It's just a bit difficult to estimate the max num of versions out in the wild, but I agree that it should work fine, also since we're basically doing the same in the service 👍 

> The impact should be minimal as only the final delta is added. [...]

Right 🚀

File: api/src/controllers/versions.ts [hunk 1]
```diff
@@ -211,9 +211,15 @@ router.get(
...
-		const saves = await service.getVersionSavesById(version['id']);
+		let current;
 
-		const current = assign({}, ...saves);
+		if (version['delta']) {
+			current = version['delta'];
+		} else {
+			const saves = await service.getVersionSavesById(version['id']);
+
+			current = assign({}, ...saves);
```
Comment (br41nslug) [inline api/src/controllers/versions.ts:221]: Should we consider deleting the old versions once they have been migrated and saved here in the fallback?

File: api/src/services/versions.ts [hunk 2]
```diff
@@ -253,6 +258,18 @@ export class VersionsService extends ItemsService {
 			delta: revisionDelta,
 		});
 
+		let existingDelta = version['delta'];
+
+		if (!existingDelta) {
+			const saves = await this.getVersionSavesById(key);
+
+			existingDelta = assign({}, ...saves);
```
Comment (br41nslug) [inline api/src/services/versions.ts:266]: Same question as above

File: api/src/services/versions.ts [hunk 3]
```diff
@@ -276,9 +293,15 @@ export class VersionsService extends ItemsService {
...
-		const saves = await this.getVersionSavesById(id);
+		let versionResult;
 
-		const versionResult = assign({}, ...saves);
+		if (delta) {
+			versionResult = delta;
+		} else {
+			const saves = await this.getVersionSavesById(id);
+
+			versionResult = assign({}, ...saves);
```
Comment (br41nslug) [inline api/src/services/versions.ts:303]: Same question as above

**Review by br41nslug — APPROVED** [review]
So far working well in my tests versions get properly migrated and seem to function as expected/before in my tests.

Just this one question about the fallback but thats not a blocker as it should already be an edge case to get there.

File: api/src/controllers/versions.ts (hunk 1 above)
Comment (licitdev) [reply api/src/controllers/versions.ts:221]: The old versions are `revisions` and they are there as a log.
